### PR TITLE
architecture: add document on ztunnel lifecycle

### PR DESCRIPTION
**Pull request by howardjohn**

### CODE REVIEW

File: architecture/ambient/lifecycle.md
Comment (howardjohn) [inline architecture/ambient/lifecycle.md:83]: Note: there is no need, but at the time of writing we do drain. I willwork on fixing it.

File: architecture/ambient/lifecycle.md
Comment (MorrisLaw) [inline architecture/ambient/lifecycle.md:27]: grammar nit:
```suggestion
This includes setting up networking rules within both the pod network namespace and the host network.
```

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:1]: IMO we should do one of: 

- put this in `ztunnel.md`
- name it `ztunnel-lifecycle.md`
- name it `traffic-capture-lifecycle.md`

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:24]: Should we just put some/all of this stuff in https://istio.io/latest/docs/ambient/architecture/traffic-redirection/ ?

I feel like most of the generic `lifecycle` stuff and calling out of specific security considerations and our mitigations of them would be useful to have there.

e.g. "It is security-critical to handle pod startup gaps for this reason, and we do in this way" probably belongs in `istio.io` docs.

I don't want to document the CNI node agent in too many disparate places, we already have

- https://istio.io/latest/docs/ambient/architecture/traffic-redirection/ (ambient-specific generic traffic capture architecture)

- https://istio.io/latest/docs/setup/additional-setup/cni/ (dataplane-mode agnostic install guide)
  
cc @craigbox for thoughts.

File: architecture/ambient/lifecycle.md
Comment (bleggett) [reply architecture/ambient/lifecycle.md:83]: xref: https://github.com/istio/ztunnel/issues/899

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:115]:
```suggestion
This enables multiple processes with the same effective UID to bind to the same port.
```

File: architecture/ambient/lifecycle.md
Comment (howardjohn) [reply architecture/ambient/lifecycle.md:24]: I agree some probably should be in public docs but there is serious value to having developer docs like this that is not tied up in public docs personas

Comment (costinm) [reply architecture/ambient/lifecycle.md:24]: I would go one step further - public docs should not be concerned with implementation details (that may change or depend on a particular implementation).

What is relevant for public docs is "The ambient L4 layer guarantees that network is secure ... from start to end"

File: architecture/ambient/lifecycle.md
Comment (costinm) [inline architecture/ambient/lifecycle.md:25]: Technically Ambient is not restricted to K8S - I would say 'the execution environment' - unless you change the title to "Ztunnel per-node daemon set" lifecycle ( which is probably better, lifecycle on a VM or as a container are different enough ).

File: architecture/ambient/lifecycle.md
Comment (costinm) [inline architecture/ambient/lifecycle.md:26]: K8S will call the Istio CNI binary (path = ??? ) , which will use a UDS socket ( path = ???) to communicate the pod info to the server running in the Istio CNI Daemon. 

The Istio CNI daemon will keep an in-memory cache with all the pods and their corresponding network namespaces.

File: architecture/ambient/lifecycle.md
Comment (costinm) [inline architecture/ambient/lifecycle.md:27]: Commenting as I read - do we have some examples with what is set on the host network - and what is set on the pod network ? The first part in particular - since it's different from normal Istio where we don't set anything on the host.

File: architecture/ambient/lifecycle.md
Comment (costinm) [inline architecture/ambient/lifecycle.md:29]: Enabled after which startup ? How can that happen ( the CNI binary will be always called).
You mean 'if the CNI binary is installed later ' ?

File: architecture/ambient/lifecycle.md
Comment (costinm) [inline architecture/ambient/lifecycle.md:31]: This deserves a lot more details - for developers and admins - it is not even close to the same setup. 

Watching the pods doesn't provide the network namespace - we perform a duct-taped search in /proc with some regex to guess which processes are pods and what are their namespaces, hoping the CRI layer is one that happens to generate ( and guarantee) a particular naming pattern that was reverse engineered from undocumented and unsupported behaviors of common CRIs.

I think it is critical for any dev ( or operator ) to understand this part - it's one of the most tricky areas.

File: architecture/ambient/lifecycle.md
Comment (costinm) [inline architecture/ambient/lifecycle.md:52]: What are the error cases ? What happens if ztunnel pod is not deployed ? 

My understanding is that CNI pod will keep in memory a list with all the network namespaces, and send them to ztunnel if it shows up ( and if CNI daemon is restarted - will do the /proc scanning to guess )

File: architecture/ambient/lifecycle.md
Comment (costinm) [inline architecture/ambient/lifecycle.md:68]: For example using persistent sessions or pod-to-pod communication will keep sending new connections after shutdown has started.

**Review by costinm — COMMENTED** [review]
Looks good, but would be better to be a bit more accurate...

File: architecture/ambient/lifecycle.md
Comment (howardjohn) [reply architecture/ambient/lifecycle.md:25]: yeah I think this is "ztunnel as a daemon set" architecture. I'll update

**Comment by costinm** [comment]
BTW - I was looking at the CNI code a bit, we have a LOT of complexity and
back and forth between ztunnel, the CNI 'server' and the CNI binary ( and
the 2 daemon sets).

Can you add a section on 'why' - only reason I can imagine ( besides
'historical reasons' ) is that CNI does host mounts while ztunnel doesn't.
And the host
mounts are mostly for the /proc mounting so we can do scans ( the installer
is independent of ztunnel and the rest).

The trickiest part around restart (IMO) is the in-memory map of container
-> network namespace, which is lost of CNI server is restarted ( perhaps -
ztunnel still has its map).

If the host 'cni plugin' had a way to save a file with the current state (
should be max 1M) and directly send it to ztunnel - a lot will be simpler
and safer...

File: architecture/ambient/lifecycle.md
Comment (bleggett) [reply architecture/ambient/lifecycle.md:24]: Implementation details should not be covered in public docs - agreed.

But I don't think "K8S-specific pod lifecycle security considerations" are component-level implementation details. That's the whole reason/justification we use a CNI plugin, and have repair for sidecars, etc etc. It's worth explaining that at a high level.

We have already had people ask about these basic architectural justifications, is why I'm suggesting we have something in public docs.

Maybe what's already there is sufficient - that's fine. Wanna make sure it is.

**Comment by costinm** [comment]
Also - in your section on requirements and startup state - it is not only
that Ztunnel must make sure iptables and listeners are set before the first
init container starts, but
also that any Authz or policies are received from Istiod before any
connection is accepted.

I don't know if we have any checks around 'host CNI plugin is blocked until
Ztunnel confirms it has at least the inbound config from XDS', or if we
have few
bugs when pods start if Istiod is not available ( which may impact security
).  If we do block all connections until policy is received - it would be
good to
mention it somewhere in the doc ( i.e. 'host cni waits for ztunnel to setup
the listeners and confirm it has all it needs from XDS')

File: architecture/ambient/lifecycle.md
Comment (bleggett) [reply architecture/ambient/lifecycle.md:1]: Also we have:

`TODO: fill in implementation details of how redirection is actually implemented.`

in `ztunnel.md` -> that TODO should probably just link to this doc.

**Comment by costinm** [comment]
I don't think we have any different requirements from what K8S CNI
requires.

All CNI implementations are required to provide a stable ( and secure )
environment before the first init container is run until all processes are
killed. We can link to the K8S docs in our public docs.
How it is done (right now) is probably mostly relevant for developers.

File: architecture/ambient/lifecycle.md
Comment (bleggett) [reply architecture/ambient/lifecycle.md:1]: Also - we should probably _remove_ some of the text in https://github.com/istio/istio/tree/master/cni/README.MD in favor of this - there's some duplication, and the former is slightly out of date.

File: architecture/ambient/lifecycle.md
Comment (bleggett) [reply architecture/ambient/lifecycle.md:27]: Some of that (should be/was) in https://github.com/istio/istio/tree/master/cni/README.MD

All that is created on the host are a single iptables rule and ipset for host probes.

That does raise a good question - if this is `/architecture/ambient` - should it have that impl detail in it?

Or should that impl detail go in `cni/README.MD`, next to the thing that actually does the impl?

File: architecture/ambient/lifecycle.md
Comment (bleggett) [reply architecture/ambient/lifecycle.md:24]: > I would go one step further - public docs should not be concerned with implementation details (that may change or depend on a particular implementation).

I don't think we can get away with that, since the entire reason we have a node agent with a CNI plugin has to do with K8S pod lifecycle implementation details.

**Comment by bleggett** [comment]
> BTW - I was looking at the CNI code a bit, we have a LOT of complexity and back and forth between ztunnel, the CNI 'server' and the CNI binary ( and the 2 daemon sets). Can you add a section on 'why' - only reason I can imagine ( besides 'historical reasons' ) is that CNI does host mounts while ztunnel doesn't. 

Most of it boils down to 

- `ztunnel` shouldn't/doesn't need a K8S client/hard dep on K8S API server.
- installing and managing the CNI plugin on the node requires host FS access and some amount of node-level machinery, and we need a plugin.
- putting all of that complexity into ztunnel would make ztunnel much larger, more complex, and more privileged - you still need the complexity, you've just moved it.

File: architecture/ambient/lifecycle.md
Comment (howardjohn) [reply architecture/ambient/lifecycle.md:24]: Implementation details in docs are good for a few reasons:
* You did something great and need to prove it's real (we provide secure networking with X properties,and here is the details if you really want)
* You did something terrible and need to explain it (sidecar init has issues so we hacked on a repair controller)
* Understanding details is useful for operations (arguably, this is the same as "you did something terrible", because making users care about low level details means we did a bad job making it operational)

I think we have a bit of each here (hopefully less terrible stuff)

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:25]:
```suggestion
The container runtime is responsible for invoking this when a Pod is being started (before the containers run).
```

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:45]:
```suggestion
During new Pod startup, we MUST meet the constraint that the network is ready the instant the application starts (including `initContainers`!).
```

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:51]:
```suggestion
The CNI plugin is executed synchronously during the creation of the Pod sandbox by the container runtime, before processes or containers are scheduled in the Pod. This ensures we can setup the network configuration and start Ztunnel.
```

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:52]:
```suggestion
Only once those two conditions are met does the CNI plugin return success - effectively allowing our plugin to block Pod scheduling until we are ready.

If the CNI plugin does not successfully complete, the Pod sandbox will fail to complete, and the Pod will not be scheduled by Kubernetes.
```

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:64]:
```suggestion
During Pod shutdown, we MUST meet the constraint that Pod traffic will be handled until all containerized processes within the Pod have terminated. We do this by waiting for K8S to tell us that the Pod has terminated, as it is the final/best/last arbiter for this.
```

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:67]: Most of this can be rolled into the statement "we don't remove until we get a pod terminated state from K8S, K8S is the arbiter"

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:85]:
```suggestion
Notably, this prevents us from sending a GOAWAY to notify peers we have shutdown, even in the cases where we could in theory do so (i.e. clients understand and respect GOAWAY)
```

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:88]:
```suggestion
```

This probably just belongs in an issue where the problem can be discussed, not this doc - the hook (or lack thereof) isn't really the problem here AFAIK, it's that you can't GOAWAY clients that aren't aware they're being proxied.

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:110]:
```suggestion
Depending on the application, this may have a range of impacts (terminating in-flight transactions, etc). In general, applications should never assume the network is 100% reliable and that no connections should be disrupted, or they will generally have[ stability problems in the context of any distributed system](https://en.wikipedia.org/wiki/Fallacies_of_distributed_computing)
```

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:114]:
```suggestion
The key to this "handoff" between Ztunnels is `SO_REUSEPORT`, which Ztunnel by default sets on all of its listeners.
```

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:116]:
```suggestion
Linux will then send new connections to one of the processes.
```

Which one? Does Linux actually do this, or is it just a best-effort case of "first-acceptor-wins"?

File: architecture/ambient/lifecycle.md
Comment (bleggett) [inline architecture/ambient/lifecycle.md:124]: This isn't a true statement in all cases, it's highly dependent on your rolling update strategy and how you configured the K8S daemonset in general - worth pointing out, even though we do hardcode that currently.

It is, strictly speaking, something that K8S orchestrates, not us.

File: architecture/ambient/lifecycle.md
Comment (bleggett) [reply architecture/ambient/lifecycle.md:24]: Yeah - I'm more concerned that we have those implementation details here _and_ elsewhere too, and I'd rather be parsimonious about duplicating those details in different spots.

Putting them here is fine - if we do a pass to remove those same details from the existing spots, if and where they exist.

Comment (craigbox) [reply architecture/ambient/lifecycle.md:24]: I agree that this is content for two separate audiences and it's OK to have some overlap.
Once this is in, we can easily mine the best bits and see what should go into the user-facing documentation.

File: architecture/ambient/lifecycle.md
Comment (craigbox) [reply architecture/ambient/lifecycle.md:25]: still holding out hope for Mesos I see

Comment (bleggett) [reply architecture/ambient/lifecycle.md:25]: obiwan.gif

File: architecture/ambient/lifecycle.md
Comment (bleggett) [reply architecture/ambient/lifecycle.md:24]: Then let's at least link to this from the `istio-cni` README then, since it already has a TODO stub for it.

Comment (bleggett) [reply architecture/ambient/lifecycle.md:24]: (and scrub a lot of the content from that README too, actually - it mostly restates a subset of what's here now)

**Comment by costinm** [comment]
On Mon, Jul 1, 2024 at 2:13 PM Ben Leggett ***@***.***> wrote:

> BTW - I was looking at the CNI code a bit, we have a LOT of complexity and
> back and forth between ztunnel, the CNI 'server' and the CNI binary ( and
> the 2 daemon sets). Can you add a section on 'why' - only reason I can
> imagine ( besides 'historical reasons' ) is that CNI does host mounts while
> ztunnel doesn't.
>
> Most of it boils down to
>
>    - ztunnel shouldn't/doesn't need a K8S client/hard dep on K8S API
>    server.
>    - installing and managing the CNI plugin on the node requires host FS
>    access and some amount of node-level machinery, and we need a plugin.
>    - putting all of that complexity into ztunnel would make ztunnel much
>    larger, more complex, and more privileged - you still need the complexity,
>    you've just moved it.
>
> That should be documented - it is a trade off, of course the arguments for
'microservices' is well known, but we need to
have it written down ( and of course other implementations of ambient may
make other choices ).
AFAIK it is common for other CNIs to run on the host ( installed as part of
the node ) and to have connections to
either K8S or a control plane.

ZTunnel has a connection to XDS - and without getting the Authz rules it
can't operate a pod network, so
there is no 'what if XDS is down' reason. And AFAIK nothing that K8S
returns can't be provided by WDS, so
probably 'we can't connect to K8S' is not a major concern ( and we should
mention in the doc that we choose to
get the data from K8S in CNI and perhaps some small notes on what data we
actually use).

The permission and separation are good reasons - at least for the 'install
ambient using daemon sets' - but
I suspect some vendors will just install at least part of it on the node,
and then it's no longer that relevant, host
filesystem access is no longer needed ( except for the /proc scanning hack
- but I think there are other options
for that with a more strict host install).

**Comment by costinm** [comment]
BTW - my main point is that it will be VERY helpful if the docs explained
this kind of details, it is far
from well understood by devs and maintainers - we may disagree on some
areas, but having the info
available, including the tradeoffs we make and the weak points is pretty
important.

File: architecture/ambient/lifecycle.md
Comment (howardjohn) [reply architecture/ambient/lifecycle.md:27]: I linked to CNI/readme.dm

File: architecture/ambient/lifecycle.md
Comment (howardjohn) [reply architecture/ambient/lifecycle.md:29]: Its if the namespace/pod is enrolled after it starts. By label, etc

File: architecture/ambient/lifecycle.md
Comment (howardjohn) [reply architecture/ambient/lifecycle.md:52]: It blocks the pod startup. I updated the doc

File: architecture/ambient/lifecycle.md
Comment (howardjohn) [reply architecture/ambient/lifecycle.md:88]: No, that is not really the problem I am trying to convey. The problem is HBONE clients, which we know can GOAWAY, cannot because we cannot send it at the right time. This isn't talking about the fact we are handling opaque TCP streams.

I will remove this CNI DEL part though, agree with your comment

**Comment by howardjohn** [comment]
Can we merge this? I am kind of surprised by the massive amount of pushback I get everytime I try to write documentation for developers in architecture/ - more so than on any of the PRs actually adding the things I am trying to retroactively document...

**Comment by craigbox** [comment]
FWIW I am in favour of treating docs like a wiki and updating early and often. Merge things and make changes rather than holding too long.

(If you want to update this for to be owned by docs I could help further)

**Comment by keithmattix** [comment]
+1 to Craig; i'm in favor of merging these kinds of these early and often

**Comment by howardjohn** [comment]
Me too, so please approve it :stuck_out_tongue:

**Review by keithmattix — APPROVED** [review]
Oops I thought I already had 😅 you'll still need a pilot and environments approver though

**Comment by bleggett** [comment]
> Can we merge this? I am kind of surprised by the massive amount of pushback I get everytime I try to write documentation for developers in architecture/ - more so than on any of the PRs actually adding the things I am trying to retroactively document...

In general, at least from my perspective - that's because 

- those kinds of docs go out of sync frequently
- have historically contained useful information that never makes its way to `istio.io`
- or contain the same information, worded slightly differently, leading to two different sources of doc entropy

When it comes to adding documentation sources of truth, being parsimonious is usually more sustainable.

**Comment by bleggett** [comment]
> I agree with treating docs like a wiki, but I don't want two wikis :D

(I am reminded here that we might have more like 3 wikis in practice now, I forgot about https://github.com/istio/istio/wiki)

**Comment by howardjohn** [comment]
If we are not going to approve dev docs, there are two outcomes:
* There are no dev docs
* There are 4+ places with dev docs, since each dev writes them in some other arbitrary places that doesn't take 3 weeks to approve.

**Review by bleggett — APPROVED** [review]

**Comment by bleggett** [comment]
> If we are not going to approve dev docs, there are two outcomes:
> 
>     * There are no dev docs
> 
>     * There are 4+ places with dev docs, since each dev writes them in some other arbitrary places that doesn't take 3 weeks to approve.

Yes - I wasn't intending to hold this up, I was mostly making a more general point - apologies, missed this needed a `ztunnel` maintainer stamp.

**Pull request merged**
